lock-api/src/main/java/com/palantir/lock/client/PartitioningTimelockService.java

```diff
@@ -0,0 +1,206 @@
+                UUID::randomUUID,
+                LOCK_BATCH_THRESHOLD,
+                UNLOCK_BATCH_THRESHOLD);
+    }
+
+    @Override
+    public TimelockService delegate() {
+        return timelockService;
+    }
+
+    @Override
+    public Set<LockToken> unlock(Set<LockToken> unenrichedTokens) {
+        Map<LockToken, LockToken> tokens = replaceProxies(unenrichedTokens);
+        if (tokens.size() <= unlockBatchThreshold) {
+            return timelockService.unlock(tokens.keySet());
+        }
+        Set<LockToken> unsuccessfulTokens = Streams.stream(Iterables.partition(tokens.keySet(), unlockBatchThreshold))
+                .map(ImmutableSet::copyOf)
+                .map(t -> Sets.difference(t, timelockService.unlock(t)))
+                .flatMap(Collection::stream)
+                .map(tokens::get)
+                .collect(Collectors.toSet());
+        return Sets.difference(ImmutableSet.copyOf(tokens.values()), unsuccessfulTokens);
+    }
+
+    @Override
+    public void tryUnlock(Set<LockToken> unenrichedTokens) {
+        Set<LockToken> tokens = replaceProxies(unenrichedTokens).keySet();
+        Streams.stream(Iterables.partition(tokens, unlockBatchThreshold))
+                .map(ImmutableSet::copyOf)
+                .forEach(timelockService::tryUnlock);
+    }
+
+    @Override
+    public LockResponse lock(LockRequest request) {
+        if (request.getLockDescriptors().size() <= lockBatchThreshold) {
+            return timelockService.lock(request);
+        }
+
+        Instant startTime = clock.instant();
+        Duration initialAcquireTimeout = Duration.ofMillis(request.getAcquireTimeoutMs());
+        List<LockDescriptor> sortedLockDescriptors =
+                Ordering.natural().immutableSortedCopy(request.getLockDescriptors());
+        BatchedLockDescriptors descriptors = new BatchedLockDescriptors();
+        try {
+            for (List<LockDescriptor> batch : Iterables.partition(sortedLockDescriptors, lockBatchThreshold)) {
+                Optional<Duration> acquireTimeout = acquireTimeout(startTime, initialAcquireTimeout);
+                if (!acquireTimeout.isPresent()) {
+                    return descriptors.failure(LockResponse.timedOut());
+                }
+
+                LockRequest batchedRequest = ImmutableLockRequest.builder()
+                        .requestId(requestIdSupplier.get())
+                        .acquireTimeoutMs(acquireTimeout.get().toMillis())
+                        .lockDescriptors(batch)
+                        .build();
+                LockResponse response = timelockService.lock(batchedRequest);
+                if (!response.wasSuccessful()) {
+                    return descriptors.failure(response);
+                }
+                descriptors.add(response.getToken());
+            }
+        } catch (RuntimeException e) {
+            throw descriptors.failure(e);
+        }
+        return descriptors.success();
```

Comment:
It's possible I'm being overly pedantic/worrying too much about the semantics here. My understanding of the guarantees of the lock service was that if you get a successful response, then there existed a point in time (not necessarily when the method returns) where you had all the locks. Currently this is true on the timelock server side, if I read the code correctly (notice that while a request is in flight, `HeldLocksCollection#shouldRemove()` will never be true).

Here if we have 2 batches and a long GC between batch 1 and 2, we might lose the locks in batch 1 which someone else might acquire but still return success, meaning we never held all of them simultaneously. I don't think this affects correctness of the Atlas protocol in that we don't commit unless we refresh our token successfully later on. 

I'm not sure how/if this will affect broken workflows where people rely on the transaction task in a `runTaskWithRetry(...)` not running without being able to acquire all the locks for an instant, though it's arguable (and I'd agree) that that's not something we should concern ourselves with. 

That said, I'm not sure I have a better solution - one way is to immediately refresh the tokens and confirm all are there before returning OK, but that's unperformant for Atlas write transactions which are going to check these tokens later anyway. Alternatively, if there's a way to clarify that this `TimelockService` might not exactly have the same semantics as a non-batching one, that would work for me too.

@sandorw for a sanity check on reasoning above?